app/controllers/users/phone_setup_controller.rb

```diff
@@ -135,5 +137,39 @@ def new_phone_form_params
         :recaptcha_mock_score,
       )
     end
+
+    def check_phone_submission_limit
+      if exceeded_phone_submission_limit?
+        return handle_too_many_phone_submissions
+      end
+      phone_submission_rate_limiter.increment
+      if exceeded_phone_submission_limit?
+        return handle_too_many_phone_submissions
+      end
+    end
+
+    def exceeded_phone_submission_limit?
+      if phone_submission_rate_limiter.exceeded_otp_send_limit?
+        return phone_submission_rate_limiter.lock_out_user
+      end
+    end
+
+    def phone_submission_rate_limiter
+      @phone_submission_rate_limiter ||= OtpRateLimiter.new(
+        phone: new_phone_form_params[:phone],
+        user: current_user,
+        phone_confirmed: UserSessionContext.authentication_or_reauthentication_context?(context),
+        limit_type: :phone_submissions,
```

Comment:
The term "submissions" feels a bit generic and I'd worry could be ambiguous (e.g. I _submit_ a code when authenticating). Elsewhere, we use terms "confirmation" and "authentication" to differentiate between setting up a phone vs. authenticating with it. Maybe we should use the confirmation terminology here?

Reply:
That's renamed to phone_fingerprint_confirmations